### PR TITLE
WAR by avoid querying device before env mutation

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #107301

We should probably fix https://github.com/pytorch/pytorch/issues/107300
properly but this works around the problem

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @aakhundov

### CODE REVIEW

**Review by bdhirsh — APPROVED** [review]

**Review by H-Huang — APPROVED** [review]
Thanks for fixing

**Review by albanD — APPROVED** [review]
Sounds ok, maybe we can avoid having this again?

File: benchmarks/dynamo/common.py [hunk 1]
```diff
@@ -3002,9 +3002,12 @@ def main(runner, original_dir=None):
                 f"--diff-branch: current branch is same as {args.diff_branch} branch, what are you diffing?"
             )
 
-    device_count = torch.cuda.device_count()
     args.use_distributed = (args.ddp or args.fsdp) and args.only
     if args.multiprocess:
+        # NB: Do NOT query device count before CUDA initialization; we're
+        # going to overwrite CUDA_VISIBLE_DEVICES and this will result in
+        # https://github.com/pytorch/pytorch/issues/107300
+        device_count = torch.cuda.device_count()
```
[the remark this answers is not in the document]
Comment (ezyang) [reply benchmarks/dynamo/common.py:3010]: I need to think about what the durable fix is. Ideally it is OK to query device count and then set the env var, as long as it happens before CUDA initialization. But that means I need to some how invalidate the device count cache. But there is no obvious hook point for setting the env var. Maybe CUDA initialization should always clear the device count??

[the remark this answers is not in the document]
Comment (albanD) [reply benchmarks/dynamo/common.py:3010]: I think the best solution here is to disable device_count() (and friends) caching if the cuda context is not initialized.

The rationale for this is that as long as the cuda context is not intialized, any change to the env variable will be picked up and so we want to properly reflect it.
But once the context is initialized the env variable will be ignored and it is ok to keep the value that was "frozen" when that happened.

[the remark this answers is not in the document]
Comment (ezyang) [reply benchmarks/dynamo/common.py:3010]: This sounds very BC breaking!

**Pull request closed without merging**